src/Simulation/Microsoft.Quantum.TargetGeneration/GenerateTarget.fs

```diff
@@ -0,0 +1,398 @@
+                let rec buildVariableName = function
+                    | QsTupleItem  one -> one.VariableName |> name
+                    | QsTuple     many -> "(" + (many |> Seq.map buildVariableName |> String.concat ",") + ")"
+                let rec buildVariableNameList accum a =
+                    match a with
+                    | QsTupleItem  one -> ((one.VariableName |> name |> ``ident``) :: accum) |> List.rev
+                    | QsTuple     many -> many |> Seq.fold (fun s i -> buildVariableNameList s i) accum
+                match spArgs with
+                | QsTupleItem one -> (one.VariableName |> name, [], [one.VariableName |> name |> ``ident``])
+                | QsTuple many    ->
+                    if many.Length = 0 then
+                        ("__in__", [], [])
+                    elif many.Length = 1 then
+                        ((buildVariableName many.[0]), 
+                         [], 
+                         [``ident`` (buildVariableName many.[0])])
+                    else
+                        ("__in__",
+                         [ ``var`` (buildVariableName spArgs) (``:=`` <| inData) ], 
+                         buildVariableNameList [] spArgs)
+            let sim = ``((`` (``as`` targetClass ((``ident`` "this" ) <|.|> (``ident`` "Factory"))) ``))``
+            let methodName = implMethodName sp |> ``ident``
+            let call = sim <.> (methodName, argNames)
+            let ret =
+                match returnType.Resolution with
+                | QsTypeKind.UnitType ->
+                    [
+                        call |> statement
+                        ``return`` ( Some ((``ident`` "QVoid") <|.|> (``ident`` "Instance")) )
+                    ]
+                | _ ->
+                    [
+                        ``return`` ( Some call )
+                    ]
+            let body = ``() => {}`` [ argName ] (argsInit @ ret) :> ExpressionSyntax
+            Some (body |> propertyFromBody)
+        | Generated SelfInverse ->
+            let adjointedBodyName =
+                match sp.Kind with
+                | QsAdjoint           -> "Body"
+                | QsControlledAdjoint -> "ControlledBody"
+//TODO: diagnostics.
+                | _ -> "Body"
+            Some (adjointedBodyName |> ``ident`` |> propertyFromBody)
+        | _ ->
+            None
+
+    let buildIntrinsicClass context targetClass op =
+        let (name, nonGenericName) = SimulationCode.findClassName context op
+        let inType   = op.Signature.ArgumentType |> SimulationCode.roslynTypeName context
+        let outType  = op.Signature.ReturnType   |> SimulationCode.roslynTypeName context
+        let constructors = [ (SimulationCode.buildConstructor context name) ]
+        let nameProperties = [ SimulationCode.buildName name; 
+                               SimulationCode.buildFullName op.FullName ]
+        let baseOp =
```

Comment:
This function could also be split out in the C# generation and used from there. And actually, I think so can a lot of things here.

Reply:
Done.